rclpy/src/rclpy/_rclpy.c

```diff
@@ -1599,6 +1599,68 @@ rclpy_publish(PyObject * Py_UNUSED(self), PyObject * args)
   Py_RETURN_NONE;
 }
 
+/// Publish a serialized message
+/**
+ * Raises ValueError if pypublisher is not a publisher capsule
+ * Raises RuntimeError if the message cannot be published
+ *
+ * \param[in] pypublisher Capsule pointing to the publisher
+ * \param[in] pymsg serialized message to send
+ * \return NULL
+ */
+static PyObject *
+rclpy_publish_raw(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pypublisher;
+  PyBytesObject * pymsg;
+  char * message;
+  Py_ssize_t pymsg_size;
+
+  if (!PyArg_ParseTuple(args, "OS", &pypublisher, &pymsg)) {
+    return NULL;
+  }
+
+  PyBytes_AsStringAndSize(pymsg, &message, &pymsg_size);
+
+  rcl_serialized_message_t msg = rmw_get_zero_initialized_serialized_message();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_ret_t ret = rmw_serialized_message_init(&msg, (size_t)pymsg_size, &allocator);
+
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(
+      RCLError,
+      "Failed to initialize message: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  msg.buffer = message;
```

Comment:
This is the line that is related to the segmentation fault. We are copying a pointer to the serialized message here, and later freeing that data when we call `rmw_serialized_message_fini`. One solution is copy the data here (not just the pointer). But, I think we can avoid the copy by initializing the serialized message struct on the stack (ie. don't call `rmw_serialized_message_init` or `rmw_serialized_message_fini`), like I've done previously here:

https://github.com/ros2/rclpy/blob/211da41d6d89e8fe7076d259909b7fef30787bff/rclpy/src/rclpy/_rclpy.c#L5053-L5058

Reply:
Ah right, that makes sense. I'll update the code and rebase the commits